### PR TITLE
Add planned enhancements to DG

### LINKED ISSUE #66: Static date for the release dates for the dorm application result
Should we have a function to change it every year? Or we assume that the dates are same for each year?

### LINKED ISSUE #153: [PE-D][Tester B] Name field parses incorrectly for names with " d/o" inside
Steps taken to reproduce: Run `add n/alice d/o muthu p/98765432`.

Expected: Person with name "alice d/o muthu" and phone number 98765432 should be added.

Actual: Person with name "alice" is added, with phone number 98765432 and dorm tag "o mutho". This is due to the " d/o" section being mistakenly parsed as dorm, instead of parsing ir as actually part of her name (d/o stands for daughter of).

Screenshots:

Suggestion:
Perhaps you can consider specifying in the UG that this is a special case that should be taken note of, or changing the prefix to something else like "dr/" :)

-------------
Labels: `severity.Medium` `type.FunctionalityBug`
original: appleraincoat/ped#3

### LINKED ISSUE #155: [PE-D][Tester A] Unable to add edit room number
In the user guide it was stated that the block and room number must be at least 2 characters long, but as a resident of RVRC our blocks are all single letters, e.g.(A, B, C ...). It would be nice if the user would be able to add block numbers of 1 character length.

-------------
Labels: `type.FeatureFlaw` `severity.Low`
original: CJerrong/ped#3

### LINKED ISSUE #158: [PE-D][Tester B] Little help is provided when user enters an incorrect input
Steps taken: Run `add n/alice d/pgpr`

Expected: Error message telling me that I am missing the phone field specifically 

Actual: Error message saying incorrect format, with example and format listed out.

Screenshot:

Suggestion/Reason:
Having would be easier for the user to understand what is wrong, especially if he is inputting many optional fields and happen to miss the phone. Saves the user some time by not having to search one by one in his original input.

-------------
Labels: `type.FeatureFlaw` `severity.Low`
original: appleraincoat/ped#4

### LINKED ISSUE #159: [PE-D][Tester C] Able to Add Overlapping Instances of Free Time
Able to add Mon:0600-0800, but it appears as another time block in the contact display:

Would be better if the timings would combine if it overlaps.

-------------
Labels: `type.FunctionalityBug` `severity.Medium`
original: chuahjiajie/ped#5

### LINKED ISSUE #164: [PE-D][Tester C] Unable to Add Free Time Across Different Days
As stated in the title, would be more convenient if you could add free time for a person from let's say Monday to Wednesday 

-------------
Labels: `type.FeatureFlaw` `severity.Low`
original: chuahjiajie/ped#7

### LINKED ISSUE #167: [PE-D][Tester C] No All Day Option
Would be nice to have an all day option like in google calendars when adding free time for a person.

-------------
Labels: `type.FeatureFlaw` `severity.Low`
original: chuahjiajie/ped#8

### LINKED ISSUE #178: [PE-D][Tester A] Might be good to have whoisfree command be able to search for a range of time
For anyone in the dorm, it might be useful to ascertain that your friend would be free for the whole duration of activity that you would want to do. 

Thus it might be useful to implement `whoisfree Day:StartTime-EndTime` which would give a better picture on their availability.

-------------
Labels: `type.FeatureFlaw` `severity.Low`
original: CJerrong/ped#9

### LINKED ISSUE #180: [PE-D][Tester B] Adding Free Time for times that span multiple days require multiple freetime Tags
Steps taken: Run `addTime 3 ft/Tue:1200-2359 ft/Wed:0000-0100`.

Suggestion: Per haps you can consider making the tags by contiguous time slots instead of 1 tag per day only, such as in this case, Tue:1200-Wed:0100.

Screenshot:

-------------
Labels: `type.FeatureFlaw` `severity.Low`
original: appleraincoat/ped#8

### LINKED ISSUE #181: [PE-D][Tester A] Case sensitivity issue with freeTimeTag
Just a slight confusion with whether the `Day` component of `freeTimeTag` is case-insensitive. 

As in the error message above, it states that timestamps are case-insensitive, but the command won't pass through. Only after changing my command from `whoisfree wed:0000` to `whoisfree Wed:0000` does the command execute and return 0 persons.

-------------
Labels: `type.FunctionalityBug` `severity.Low`
original: CJerrong/ped#10

### LINKED ISSUE #190: [PE-D][Tester D] Add Free Time Feature does not account for different dates.
In the screenshot above, one can clearly see that the contact named 'Alex' has a free time period on Monday from 8am to 12pm.

However, a user using this application will not be able to tell which Monday this is referring to, which poses as quite a big issue from the end-user perspective. However, as technically users could be diligent about it and manually keep track by only adding and deleting free periods for the immediate week, I labelled this severity as medium.

However, this does not remove the hassle of remembering which week the label is referring to, and it also means that users cannot add multiple free periods occurring on the same day at the same time. (e.g. someone could be free one Monday from now at 8am to 12pm, two Mondays from now also at 8pm to 12pm etc.)

Therefore this feature flaw arises because the application is unable to distinguish duplicated free time periods due to the fact that the dates are not being stored, and thus constitutes a feature flaw for the reasons stated above.

-------------
Labels: `severity.Medium` `type.FeatureFlaw`
original: jamessinmaojun/ped#9

### LINKED ISSUE #191: [PE-D][Tester B] Help is not very useful
`help` is not very useful as it only leads to another link to the UG.

Consider changing the Help Window to display a short summary of how to use all the commands, so that the user can get comprehensive help without leaving the app

-------------
Labels: `type.FeatureFlaw` `severity.Low`
original: appleraincoat/ped#11

### LINKED ISSUE #192: [PE-D][Tester B] Phone Number may be too restrictive
Current restrictions only allow for the phone number to be exactly 8 digits and start with 8 or 9.

Consider allowing any numerical string, as those with overseas numbers may not have 8 digits. Those without phones (using landlines) will have phone numbers that start with 6.

-------------
Labels: `type.FeatureFlaw` `severity.Low`
original: appleraincoat/ped#12

### LINKED ISSUE #195: [PE-D][Tester B] Time format for addTime, deleteTime is case-sensitive (for DAY)
Steps taken: Run `add n/timothy p/99999999 ft/tue:1000-1100`.

Expected: Able to add inside (case insensitive), or have an error message saying that it is case-sensitive

Actual: Error message just listing the format, saying that the format is invalid with no reason

Screenshot: 

-------------
Labels: `type.FeatureFlaw` `severity.Low`
original: appleraincoat/ped#15

### LINKED ISSUE #196: [PE-D][Tester B] Prefixes are case sensitive
Steps taken: Run `add N/john P/87658765`
Expected: Add the contact with the specified parameters.
Actual: Parsed incorrectly as the prefixes are case-sensitive

Suggestions: Consider making the prefixes case sensitive, as it is easy to mistype capital letters for some users

-------------
Labels: `type.FeatureFlaw` `severity.Low`
original: appleraincoat/ped#16

### LINKED ISSUE #199: [PE-D][Tester D] Add Free Time Feature does not account for inconsistent day casing
If I enter the following command:
`addTime 1 ft/mon:0800-1200`

I will get an error as follows:

While I understand that this is not a bug per se, it would definitely be useful to accommodate to different name casing for the days (I can definitely foresee some users entering `mon` or `MON` instead of `Mon`, and not allowing different casings could pose as a rather common inconvenience)

Not allowing different casings is also inconsistent with the other commands such as `find`, which have parameters which are case-insensitive.

-------------
Labels: `severity.Low` `type.FeatureFlaw`
original: jamessinmaojun/ped#11

### LINKED ISSUE #201: [PE-D][Tester D] Delete Free Time Feature does not account for inconsistent day casing
Similar to the other issue on the `addTime` command,

If I enter the following command:
`deleteTime 1 ft/mon:0800-1200`

I will get an error as follows:

While I understand that this is not a bug per se, it would definitely be useful to accommodate to different name casing for the days (I can definitely foresee some users entering mon or MON instead of Mon, and not allowing different casings could pose as a rather common inconvenience)

Additionally, I understand that `addTime` and `deleteTime` are not the only commands using this style of day formatting. This feature flaw would affect other commands such as the `whoisfree` command too.

Not allowing different casings is also inconsistent with the other commands such as find, which have parameters which are case-insensitive.

-------------
Labels: `severity.Low` `type.FeatureFlaw`
original: jamessinmaojun/ped#13

### LINKED ISSUE #207: [PE-D][Tester C] Better Execution of "whoisfree"
Although not stated in the UG explicitly, I am assuming that this app is for people to quickly pull up when they are free and want to quickly make last minute plans with someone. 

If that is the case, it would be better for contacts to be sorted by free time. More specifically the person(s) that have free time closest to the current time should be displayed at the top. 

This is more free for the typical use case of the app.

-------------
Labels: `type.FeatureFlaw` `severity.Low`
original: chuahjiajie/ped#13

### LINKED ISSUE #208: [PE-D][Tester C] App Only works for 1 Week
The details in the app would be rendered redundant by the end of the week as it only accounts for days and not days or months. Either this needs to be changed to include more date details, or the free time data needs be easily clearable every week. For example, by having a "clearTime" command, to clear all the free time data for all contacts, instead of having to "clear" all contacts, or delete all the free time data one by one.

-------------
Labels: `type.FunctionalityBug` `severity.Medium`
original: chuahjiajie/ped#14

### PR DESCRIPTION
**Pull request by PallonCX**
Planned enhancements and the corresponding issues that it covers:
[1] Enable cross day free time interval (Closes #180) (Closes #164)
[2] Enable free time interval merging (Closes #159)
[4] Allow more flexibility in command prefixes (Closes #196)
[5] Supporting more forms of name (Closes #153)
[6] Supporting more forms of room number (Closes #155)
[7] Allow more flexibility in free time tag format (Closes #201) (Closes #199) (Closes #181) (Closes #195)
[8] Support more accurate free time (Closes #190) (Closes #208)
[9] Make error messages more specific (Closes #158)
[10] Make the `whoisfree` command to search for a range of time (Closes #178)
[11] Edit command can remove value for optional fields

Some issues that are excluded from planned enhancement due to low importance:
Closes #66
Closes #167
Closes #191
Closes #192
Closes #207

### CODE REVIEW

**Review by nyankoclaws — APPROVED** [review]
LGTM

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -492,3 +483,68 @@ testers are expected to do more *exploratory* testing.
...
+
+5. **Supporting more forms of name:**
+   * Currently, only name form that consist of alphanumeric characters and spaces in between is allowed.
+   * It may improve the diversity of the application, by supporting name in different forms, including names in different languages and names that contain non-alphanumeric symbols.
+
+6. **Supporting more forms of room number:**
+   * Currently, the room number format is {block}-{floor}-{room number}, where block and room number are at least 2 alphanumeric characters and floor is strictly 2 alphanumeric characters. E.g. `nw-12-12`.
+   * To accommodate more room number format, it will be good to change the minimum length of block and room number as 1.
+
+7. **Allow more flexibility in free time tag format:**
```
Comment (donwong2308) [inline docs/DeveloperGuide.md:520]: Maybe can add flexible time formats also like 11am-2pm

**Comment by PallonCX** [comment]
Closes #168.

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -492,3 +501,68 @@ testers are expected to do more *exploratory* testing.
...
+   * To improve user experience, it could be useful to merge free time tags that have overlapping time intervals.
+     * For example, `Mon:1000-1200` and `Mon:1100-1300` can be combined as `Mon:1000-1300`.
+     * For example, `Mon:2000-2359` and `Tue:0000-0800` can be combined as `Mon:2000-Tue:0800`.
+   * The merging process could be done when new free time tag is added to the person, e.g. via `add` command, `addTime` command, or `edit` command.
+
+3. **Delete free time within an interval:**
+   * Additionally, it could be useful to be able to delete free times that fall within a time interval, e.g. via `deleteTime` command.
+   * For example, an initial freeTimeTag could be `Mon:1000-1400`. Executing `deleteTime INDEX ft/Mon:1100-1200` will result in the new freeTimeTags `Mon:1000-1100` and `Mon:1200-1400`.
+
+4. **Allow more flexibility in command prefixes:**
```
Comment (donwong2308) [inline docs/DeveloperGuide.md:525]: I think need to specify Make command prefixes case-insensitive.

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -492,3 +501,68 @@ testers are expected to do more *exploratory* testing.
...
+
+5. **Supporting more forms of name:**
+   * Currently, only name form that consist of alphanumeric characters and spaces in between is allowed.
+   * It may improve the diversity of the application, by supporting name in different forms, including names in different languages and names that contain non-alphanumeric symbols.
+
+6. **Supporting more forms of room number:**
+   * Currently, the room number format is {block}-{floor}-{room number}, where block and room number are at least 2 alphanumeric characters and floor is strictly 2 alphanumeric characters. E.g. `nw-12-12`.
+   * To accommodate more room number format, it will be good to change the minimum length of block and room number as 1.
+
+7. **Allow more flexibility in free time tag format:**
```
Comment (donwong2308) [inline docs/DeveloperGuide.md:538]: Can rephrase to Make day format in free time tag case- insensitive

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -492,3 +501,68 @@ testers are expected to do more *exploratory* testing.
...
+   * Currently, free time tag only accept forms of `DDD:HHmm-HHmm`, e.g. `Mon:1200-1300`.
+   * It may provide more convenience to user by disabling case sensitivity on the day part, e.g. allowing `mon:1000-1200`, `Mon:1000-1200`, `MON:1000-1200`.
+
+8. **Support more accurate free time:**
```
Comment (donwong2308) [inline docs/DeveloperGuide.md:542]: We might need to only choose one of the solutions, so perhaps we could go with the date one but we can say that its like an extension. So we still have our original free time tags that is recurring but we can also specify a data on top of that if we need. TLDR default free time tag is recurring weekly unless we specify a date

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -492,3 +501,68 @@ testers are expected to do more *exploratory* testing.
...
+   * A possible solution is to specify actual date as part of the free time interval, e.g. `01/01:Mon:1000-1200`.
+     * Pros: It provides more customizable user experience.
+     * Cons: It requires users to have detailed schedule information of a person in their contact list, where the person may not be willing to disclose. It also requires users to provide more complicated parameters, which is more troublesome and error-prone.
+   * Another possible solution is to host the application online, and allow user to specify their own free times. 
+     * Pros: User may connect to the real-time state of another user, and able to view the real-time updated free time of others.
+     * Cons: More changes may need to be done on the application, and the developer team will need to re-evaluate the necessity of certain commands. 
+       * For example, users may not need to manually input and record the detail of others, as they only need to update their own profile, and others can view the real-time state of it. 
+       * Hence, some commands e.g. `addTime`, `deleteTime`, and `edit`, are no longer applicable to the application.
+     
+9. **Make error messages more specific:**
```
Comment (donwong2308) [inline docs/DeveloperGuide.md:553]: We need to specify which error message we targeting I think, below is the example they provided:
image

**Review by donwong2308 — COMMENTED** [review]
Left some comments for the title to be more specific and to pick 1 solution for point 8

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (PallonCX) [reply docs/DeveloperGuide.md:525]: Will change in the next update.

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (PallonCX) [reply docs/DeveloperGuide.md:538]: Will change in the next update.

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (PallonCX) [reply docs/DeveloperGuide.md:542]: Will change in the next update.

File: docs/DeveloperGuide.md (hunk 5 above)
Comment (PallonCX) [reply docs/DeveloperGuide.md:553]: Will change in the next update.

**Review by donwong2308 — APPROVED** [review]
LGTM

**Review by larainezo — APPROVED** [review]
LGTM

**Review by JustWeiHao — APPROVED** [review]
LGTM

**Pull request merged**